### ISSUE TITLE
ssl bufferevent with underlying bufferevent - infinite loop/stack overflow.

### ISSUE THREAD

**Issue by TomerHeber** [libevent/libevent#952]
We have recently upgraded from an old libevent 2.0.x to new 2.1.x libevent.
We're using ssl bufferevent with an additional underlying ssl bufferevent (ssl -> ssl tunnel).

The underlying bufferevent is released manually by our code. In other words, the flag "BEV_OPT_CLOSE_ON_FREE" is unused, instead something of this sort is executed in our code:

```
 while (m_bev)
    {
        SSL *ssl;

        if ((ssl = bufferevent_openssl_get_ssl(m_bev)) != NULL)
        {
			// SSL_shutdown...
        }

        struct bufferevent *m_underlyingBev = bufferevent_get_underlying(m_bev);
        bufferevent_free(m_bev);
        m_bev = m_underlyingBev;

        if (m_bev != NULL)
        { 
            // disable buffer event.
        }
    }
```

I suspect commit 8939676706b8e2a125c5e3344f8672ddfadeb4e1 is the root cause of the issue we're observing.
Regardless, I now wonder if the underlying bufferevent must become NULL after bufferevent_free is called (especially if "BEV_OPT_CLOSE_ON_FREE" is not used).

Observing the unlink code "bev_ssl->underlying = NULL" is commented out (why?).
Maybe it's still required in the "else" section?
**In addition, why was "bufferevent_unsuspend_read_" added to the else? (This is probably what's causing our issue).**

```
static void
be_openssl_unlink(struct bufferevent *bev)
{
	struct bufferevent_openssl *bev_ssl = upcast(bev);

	if (bev_ssl->bev.options & BEV_OPT_CLOSE_ON_FREE) {
		if (bev_ssl->underlying) {
			if (BEV_UPCAST(bev_ssl->underlying)->refcnt < 2) {
				event_warnx("BEV_OPT_CLOSE_ON_FREE set on an "
				    "bufferevent with too few references");
			} else {
				bufferevent_free(bev_ssl->underlying);
				/* We still have a reference to it, via our
				 * BIO. So we don't drop this. */
				// bev_ssl->underlying = NULL;
			}
		}
	} else {
		if (bev_ssl->underlying) {
			if (bev_ssl->underlying->errorcb == be_openssl_eventcb)
				bufferevent_setcb(bev_ssl->underlying,
				    NULL,NULL,NULL,NULL);
			bufferevent_unsuspend_read_(bev_ssl->underlying,
			    BEV_SUSPEND_FILT_READ);
		}
	}
}
```

I'm not very familiar with the libevent code. Therefore, I may be completely off. Any feedback/ideas would be appreciated.

The following is an example of an observed stack trace:

<details>

```
_#0  0x00007fc06f7601df in bufferevent_suspend_read_ () from /usr/lib64/libevent-2.1.so.6
#1  0x00007fc06f9b0c20 in stop_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#2  0x00007fc06f9b0f1c in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#3  0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#4  0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#5  0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#6  0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#7  0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#8  0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#9  0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#10 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#11 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#12 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#13 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#14 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#15 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#16 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#17 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#18 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#19 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#20 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#21 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#22 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#23 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#24 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#25 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#26 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#27 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#28 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#29 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#30 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#31 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#32 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#33 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#34 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#35 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#36 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#37 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#38 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#39 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#40 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#41 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#42 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#43 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#44 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#45 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#46 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#47 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#48 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#49 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#50 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#51 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#52 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#53 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#54 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#55 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#56 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#57 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#58 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#59 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#60 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#61 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#62 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#63 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#64 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#65 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#66 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#67 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#68 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#69 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#70 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#71 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#72 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#73 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#74 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#75 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#76 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#77 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#78 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#79 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#80 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#81 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#82 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#83 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#84 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#85 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#86 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#87 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#88 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#89 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#90 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#91 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#92 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#93 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#94 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#95 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#96 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#97 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#98 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#99 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#100 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#101 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#102 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#103 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#104 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#105 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#106 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#107 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#108 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#109 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#110 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#111 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#112 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#113 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#114 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#115 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#116 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#117 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#118 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#119 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#120 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#121 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#122 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#123 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#124 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#125 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#126 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#127 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#128 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#129 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#130 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#131 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#132 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#133 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#134 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#135 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#136 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#137 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#138 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#139 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#140 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#141 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#142 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#143 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#144 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#145 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#146 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#147 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#148 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#149 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#150 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#151 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#152 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#153 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#154 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#155 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#156 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#157 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#158 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#159 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#160 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#161 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#162 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#163 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#164 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#165 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#166 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#167 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#168 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#169 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#170 0x00007fc06f9b24fa in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#171 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#172 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#173 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#174 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#175 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#176 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#177 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#178 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#179 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#180 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#181 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#182 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#183 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#184 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#185 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#186 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#187 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#188 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#189 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#190 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#191 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#192 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#193 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#194 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#195 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#196 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#197 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#198 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#199 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#200 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#201 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#202 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#203 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#204 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#205 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#206 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#207 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#208 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#209 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#210 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#211 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#212 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#213 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#214 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#215 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#216 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#217 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#218 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#219 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#220 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#221 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#222 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#223 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#224 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#225 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#226 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#227 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#228 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#229 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#230 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#231 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#232 0x00007fc06f9b0ab2 in start_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#233 0x00007fc06f9b241e in be_openssl_enable () from /usr/lib64/libevent_openssl-2.1.so.6
#234 0x00007fc06f76033a in bufferevent_unsuspend_read_ () from /usr/lib64/libevent-2.1.so.6
#235 0x00007fc06f9b0c85 in stop_writing () from /usr/lib64/libevent_openssl-2.1.so.6
#236 0x00007fc06f9b0f28 in conn_closed () from /usr/lib64/libevent_openssl-2.1.so.6
#237 0x00007fc06f9b12bd in do_read () from /usr/lib64/libevent_openssl-2.1.so.6
#238 0x00007fc06f9b1874 in consider_reading () from /usr/lib64/libevent_openssl-2.1.so.6
#239 0x00007fc06f9b1b93 in be_openssl_readcb () from /usr/lib64/libevent_openssl-2.1.so.6
#240 0x00007fc06f7606a7 in bufferevent_run_deferred_callbacks_locked () from /usr/lib64/libevent-2.1.so.6
#241 0x00007fc06f76cc4f in event_process_active_single_queue () from /usr/lib64/libevent-2.1.so.6
#242 0x00007fc06f76d1c0 in event_process_active () from /usr/lib64/libevent-2.1.so.6
#243 0x00007fc06f76d924 in event_base_loop () from /usr/lib64/libevent-2.1.so.6
#244 0x00007fc06f76d250 in event_base_dispatch () from /usr/lib64/libevent-2.1.so.6_
```

</details>

**Comment by azat** [comment]
Hi! Thanks for the report, I will take a look, in the mean time can you provide a minimal reproducer? (this will really helps)

**Comment by TomerHeber** [comment]
Hi Azat,

I appreciate the assistance. I forgot to mention we also use the "deferred callback flag".

Unfortunately, this is not something I can easily reproduce. We have many machines in production, every hour or so a CPU core "dies" by going to 100% cpu on an infinite loop.
Normally I would say our own source code is the root cause. However, this has been stable on 2.0.x for many years.

We are going to try and remove the call to "bufferevent_unsuspend_read_(bev_ssl->underlying, BEV_SUSPEND_FILT_READ)" from the function be_openssl_unlink as I suspect this is the root cause. I will let you know the results.

I do wonder why it was added in the first place. It's not being called if BEV_OPT_CLOSE_ON_FREE is enabled. So maybe it shouldn't be there at all?

Thank you.

**Comment by TomerHeber** [comment]
After patching - no success in solving the issue.
After further reviewing the libevent code and the stacktrace, we're going to change the way/order bufferevents are released in our source code. I'll update if resolved. If that (hopefully) be the case, I'll close the issue.

**Comment by TomerHeber** [comment]
I'll describe the issue and afterwards close this "bug".
Calling bufferevent_free does not necessarily release the bufferevent.
Reason being there are pending callback (to be executed) for the bufferevent. Meaning the bufferevent will be released only after all pending callbacks have been executed.

Calling bufferevent_free(bev), SSL_free(ssl) on the underlying SSL and close(fd) on the underlying socket may cause issues (e.g. segfaults).
If there are pending callbacks libevent may try and access ssl and fd after they have been released.
I do not know if it's a design error or a bug.

I would expect libevent to omit pending callbacks after bufferevent_free has been called (to avoid this type of race conditions). Unfortunately, that is not case. 

The current work around is to wait several seconds after calling "bufferevent_free", before releasing the memory for the SSL and the socket. The alternative is to use "bev_opt_close_on_free" flag, but this does not guarantee a clean SSL shutdown.

**Comment by azat** [comment]
>Reason being there are pending callback (to be executed) for the bufferevent. Meaning the bufferevent will be released only after all pending callbacks have been executed.

Pending callbacks should not be a problem, since bufferevent_free reset callbacks.

>I appreciate the assistance. I forgot to mention we also use the "deferred callback flag".

I guess this can cause this problem, since I don't see that they are canceled.
Can you try the referenced patch? (see above)

>The alternative is to use "bev_opt_close_on_free" flag, but this does not guarantee a clean SSL shutdown.

You mean with call to `SSL_shutdown`?

**Comment by TomerHeber** [comment]
Hi Azat,

Thank you for the response.

If the deferred are not canceled (this is what I observed) this makes total sense now!

Yes, I mean SSL_shutdown. This is the primary reason we chose not to use bev_opt_close_on_free.

How about adding a new flag and applying something of this sort to be_openssl_destruct:
```
be_openssl_destruct(struct bufferevent *bev)
{
	struct bufferevent_openssl *bev_ssl = upcast(bev);

	**if (bev_ssl->bev.options & BEV_OPT_DIRTY_SSL_SHUTDOWN_ON_FREE) {
		  	int sslError = bufferevent_get_openssl_error(bev);
			if (sslError != SSL_ERROR_SYSCALL && sslError != SSL_ERROR_SSL)
			{
				SSL_set_shutdown(bev_ssl->ssl, SSL_RECEIVED_SHUTDOWN);
				SSL_shutdown(bev_ssl->ssl);
			}
	}**

	if (bev_ssl->bev.options & BEV_OPT_CLOSE_ON_FREE) {
		if (! bev_ssl->underlying) {
			evutil_socket_t fd = -1;
			BIO *bio = SSL_get_wbio(bev_ssl->ssl);
			if (bio)
				fd = BIO_get_fd(bio, NULL);
			if (fd >= 0)
				evutil_closesocket(fd);
		}
		SSL_free(bev_ssl->ssl);
	}
}
```

Not ideal but given the complexities of a proper shutdown, this additional flag will allow us to use BEV_OPT_CLOSE_ON_FREE.

About testing the patch: the relevant team has been made aware of the suggested patch. I believe they will be able to test it. Will update once I have I hear something from them.

**Comment by azat** [comment]
>How about adding a new flag and applying something of this sort to be_openssl_destruct:

It is not always enough, see https://github.com/libevent/libevent/issues/943#issuecomment-570906572

Can you check the following patch 64b90f8b6d2dcfd67f7e994b3997eb3d83f5df0e? (without any modifications in your code to avoid this issue)

**Comment by TomerHeber** [comment]
Yes, we will test. I sent you a comment about the patch (regarding ref count).
Let me know what your opinion.

**Comment by TomerHeber** [comment]
I am well aware it is not enough, and in many cases insecure.
That's why I referred to it as "dirty shutdown". It will allow us to use BEV_OPT_CLOSE_ON_FREE with at least a "dirty shutdown".

**Comment by azat** [comment]
>That's why I referred to it as "dirty shutdown". It will allow us to use BEV_OPT_CLOSE_ON_FREE with at least a "dirty shutdown".

Did not read carefully, sorry.
Yes this is also an option (and I would prefer to implement both, since I think that not lots of people care about proper shutdown, but still needs it)

**Comment by TomerHeber** [comment]
Hi Azat,

Wanted to give you quick update:
After the patch it looks like there is an improvement and the issue is now rarely observed. We are still monitoring the system and will update again soon.

Thank you,
Tomer.

**Comment by azat** [comment]
>issue is now rarely observed

But it still pops up?

**Comment by TomerHeber** [comment]
> > issue is now rarely observed
> 
> But it still pops up?

That's what I understood from the team. However, I think it's too early to call as no stack trace was taken during the incident.
